packages/flutter/lib/src/services/hardware_keyboard.dart

```diff
@@ -0,0 +1,894 @@
+///
+/// A [KeyMessageHandler] processes a [KeyMessage] and returns whether
+/// the message is considered handled. Handled messages should not be
+/// propagated to other native components.
+typedef KeyMessageHandler = bool Function(KeyMessage message);
+
+/// A singleton class that processes key messages from the platform and
+/// dispatches converted messages accordingly.
+///
+/// [KeyEventManager] receives platform key messages by [handleKeyData]
+/// and [handleRawKeyMessage], sends converted events to [HardwareKeyboard]
+/// and [RawKeyboard] for recording keeping, then dispatches the [KeyMessage]
+/// to [keyMessageHandler], the global message handler.
+///
+/// [KeyEventManager] also resolves cross-platform compatibility of keyboard
+/// implementations. Legacy platforms might have not implemented the new key
+/// data API and only sends raw key data on each key message. [KeyEventManager]
+/// recognize platform types as [KeyEventVehicle] and dispatches events in
+/// different ways accordingly.
+///
+/// [KeyEventManager] is typically created, owned, and invoked by
+/// [ServicesBinding].
+class KeyEventManager {
+  /// Create an instance.
+  ///
+  /// This is typically only called by [ServicesBinding].
+  KeyEventManager(this._hardwareKeyboard, this._rawKeyboard);
+
+  /// The global handler for all hardware key messages of Flutter.
+  ///
+  /// Key messages received from the platform is first sent to [RawKeyboard]'s
+  /// listeners and [HardwareKeyboard]'s handlers, then sent to
+  /// [keyMessageHandler], regardless of the results of [HardwareKeyboard]'s
+  /// handlers. (Handler result is explained below.) The result from the
+  /// handlers and [keyMessageHandler] are combined and returned to the plaform.
+  ///
+  /// This handler is normally set by the [FocusManager] so that it can control
+  /// the key event propagation to focused widgets. Applications that uses the
+  /// focus system (see [Focus] and [FocusManager]) to receive key events
+  /// does not need to set this field.
+  ///
+  /// ## Handler result
+  ///
+  /// Key messages on the platform are given to Flutter to be handled by the
+  /// engine. If they are not handled, then the platform will continue to
+  /// distribute the keys (i.e. propagate them) to other (possibly non-Flutter)
+  /// components in the application. The return value from this handler tells
+  /// the platform to either stop propagation (by returning true: "event
+  /// handled"), or pass the event on to other controls (false: "event not
+  /// handled").
+  ///
+  /// If you are not using the [FocusManager] to manage focus, set this
+  /// attribute with a [KeyMessageHandler] that returns true if the propagation
+  /// on the platform should not be continued. Otherwise, key events will be
+  /// assumed to not have been handled by Flutter, and will also be sent to
+  /// other (possibly non-Flutter) controls in the application.
+  ///
```

Comment:
You might mention that unhandled keys will cause the system "bonk" noise on macOS.